spec/syntax.md

```diff
@@ -684,6 +684,17 @@ except for U+0000 NULL or the surrogate code points U+D800 through U+DFFF.
 
 All code points are preserved.
 
+Two _literals_ are considered equal if they consist of the same sequence of Unicode
+code points.
```

Comment:
This seems a bit excessive. While most _name_ comparisons are internal to the spec, AFAIK the only _literal_ comparison in the spec is for duplicate variant key lists, which tbh I'd prefer to be done with normalization.

All other _literal_ value handling is done by functions, which we should not restrict from applying normalization in their internal processing.

Reply:
I agree. I'd go a bit further in encouraging function specifiers to normalize when matching: and we make the standard functions do so. That is, something along the lines of:

----

An NFC comparison (aka Unicode canonically equivalent comparison)  produces the same results as if each string value being compared were converted to the Unicode Normalization Form C (NFC). For example, with an NFC comparison against the literal |U\x{3308}|, the same result is obtained as if the literal were |\x{DC}|. For more examples, see the [Unicode Standard](https://www.unicode.org/versions/latest/core-spec/chapter-2/#G287424).

When determining whether two variant key lists are duplicates, NFC comparison MUST be used for literals.

When a selector function evaluates matches to literal keys, the matches SHOULD use NFC comparison. Moreover, the implementation of the standard selector functions MUST use NFC comparison. Thus the standard :string selector function MUST match a string input parameter of "U\x{3308}" with the literal |\x{DC}|. 

----

BTW: Some selector functions, such as the standard numeric selectors, only match literals with all ASCII characters. ASCII literals never change when converted to NFC, and there are only 3 non ASCII characters that change to ASCII. So selector functions whose literals don't include ";", "`", or "K" don't need to use NFC comparison; that includes our numeric selectors.

Char. | Code Point | Name
-- | -- | --
; | U+037E | GREEK QUESTION MARK
` | U+1FEF | GREEK VARIA
K | U+212A | KELVIN SIGN

Reply:
I think it is far, far more likely that people will make mistakes with non-NFC literals (or input) than the really, really obscure edge case of someone wanting to match non-normalized text.

Reply:
> First, do we agree that literals MAY be non-normalized/denormalized?

As with pattern text, I agree that we should not require the normalization of literal values.

> Literal comparison is [...] also for matching between the sorted results of a selector and the keys in the message (the sorting is done by a function, but not the matching after sorting).

Regarding the latter, we say this: https://github.com/unicode-org/message-format-wg/blob/80bec5215a1a3fe10abdbd75a9489f9f42d470d7/spec/formatting.md?plain=1#L511-L514

That MUST is requiring the processing to not normalise any of the values, even if it did so for its internal processing.

I'd be completely fine with us normalising the keys before they're passed to the function, or at least allowing an implementation to do so.

> > When determining whether two variant key lists are duplicates, NFC comparison MUST be used for literals.
> 
> This is the opposite of what @eemeli is saying? If we allow normalization (but don't require it) we also allow the lack of it.

I'm aligned with @macchiati here. We don't _need_ to normalize key values, but we should do their comparison when checking for duplicate key lists as if they were normalized.